cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -223,23 +228,29 @@ public void mapReduceInfo(HttpRequest request, HttpResponder responder,
       Id.Run run = new Id.Run(programId, runId);
       ApplicationSpecification appSpec = store.getApplication(programId.getApplication());
       if (appSpec == null) {
-        responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Application not found: %s",
-                                                                         programId.getApplication()));
-        return;
+        throw new NotFoundException(programId.getApplication());
       }
       if (!appSpec.getMapReduce().containsKey(mapreduceId)) {
-        responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Program not found: %s", programId));
-        return;
+        throw new NotFoundException(programId);
       }
+//      if (store.getRun(programId, runId) == null) {
+//        throw new NotFoundException(run);
+//      }
+
+
+      MRJobInfo mrJobInfo;
+      try {
+        mrJobInfo = mrJobClient.getMRJobInfo(run);
+      } catch (IOException ioe) {
+        LOG.warn("Failed to get run history for runId from JobClient: {}. Falling back to Metrics system.");
+        mrJobInfo = mapReduceMetricsInfo.getMRJobInfo(run);
```

Comment:
The outer catch block will get this. And log and respond 500.
The inner catch block is just so that we can fall back to mapReduceMetricsInfo.
Does that make sense?